charms/istio-pilot/src/istioctl.py

```diff
@@ -0,0 +1,201 @@
+        try:
+            subprocess.check_output(
+                [
+                    self._istioctl_path,
+                    "upgrade",
+                    "-y",
+                    *self._istioctl_flags,
+                ]
+            )
+        except subprocess.CalledProcessError as cpe:
+            raise IstioctlError(
+                "Upgrade failed during `istioctl upgrade` with error code" f" {cpe.returncode}"
+            ) from cpe
+
+    def version(self) -> dict:
+        """Returns istio client and control plane versions."""
+        try:
+            version_string = subprocess.check_output(
+                [
+                    self._istioctl_path,
+                    "version",
+                    f"-i={self._namespace}",
+                    "-o=yaml",
+                ]
+            )
+        except subprocess.CalledProcessError as cpe:
+            raise IstioctlError("Failed to get Istio version") from cpe
+
+        version_dict = yaml.safe_load(version_string)
+        return {
+            "client": get_client_version(version_dict),
+            "control_plane": get_control_plane_version(version_dict),
+        }
+
+
+def get_client_version(version_dict: dict) -> str:
+    """Returns the client version from a dict of `istioctl version` output
+
+    Args:
+        version_dict (dict): A dict of the version output from `istioctl version -o yaml`
+
+    Returns:
+        (str) The client version
+    """
+    try:
+        version = version_dict["clientVersion"]["version"]
+    except (KeyError, TypeError):
+        # TypeError in case version_dict is None
+        raise IstioctlError("Failed to get client version - no version found in output")
+    return version
+
+
+def get_control_plane_version(version_dict: dict) -> str:
+    """Returns the control plane version from a dict of `istioctl version` output
+
+    Args:
+        version_dict (dict): A dict of the version output from `istioctl version -o yaml`
+
+    Returns:
+        (str) The control plane version
+    """
+    # Assert that we have only one mesh and it says it is a pilot.  Not sure how we can handle
+    # multiple meshes here.
+    error_message_template = "Failed to get control plane version - {message}"
+    try:
+        meshes = version_dict["meshVersion"]
+    except KeyError:
+        raise IstioctlError(error_message_template.format(message="no control plane found"))
+
+    if len(meshes) == 0:
+        raise IstioctlError(error_message_template.format(message="no mesh found"))
+    if len(meshes) > 1:
+        raise IstioctlError(error_message_template.format(message="too many meshes found"))
+
+    mesh = meshes[0]
+
+    try:
+        if mesh["Component"] != "pilot":
+            raise IstioctlError(error_message_template.format(message="no control plane found"))
+        version = mesh["Info"]["version"]
+    except KeyError:
+        raise IstioctlError(error_message_template.format(message="no control plane found"))
```

Comment:
where is the missing key which might cause the exception? If component, then could be simplified to:
```suggestion
    if "Component" not in mesh or mesh["Component"] != "pilot":
        raise IstioctlError(error_message_template.format(message="no control plane found"))
    version = mesh["Info"]["version"]
```

Reply:
I don't know if there's cases where this data can be incomplete, so I'm trying to protect against any of these missing keys (`Component`, `Info`, or `Info.version`).  It could be collapsed, but it'll be something like:
```python
if "Component" not in mesh 
  or mesh["Component"] != "pilot" 
  or "Info" not in mesh 
  or "version" not in mesh["Info"]:
  raise IstioctlError(error_message_template.format(message="no control plane found"))
version = mesh["Info"]["version"]
```